File: question_1/task.py

class Task:
    def __init__(self, task_id, title, description, priority, deadline, completed=False):
        self.task_id = task_id
        self.title = title
        self.description = description
        self.priority = priority
        self.deadline = deadline
        self.completed = completed

    def mark_complete(self):
        self.completed = True

    def edit_task(self, title=None, description=None, priority=None, deadline=None):
        if title:
            self.title = title
        if description:
            self.description = description
        if priority:
            self.priority = priority
        if deadline:
            self.deadline = deadline

class TaskManager:
    def __init__(self):
        self.tasks = []

    def add_task(self, task):
        self.tasks.append(task)

    def edit_task(self, task_id, title=None, description=None, priority=None, deadline=None):
        for task in self.tasks:
            if task.task_id == task_id:
                task.edit_task(title, description, priority, deadline)
                return True
        return False

    def delete_task(self, task_id):
        self.tasks = [task for task in self.tasks if task.task_id != task_id]

    def get_tasks(self):
        return self.tasks

    def get_upcoming_tasks(self):
         now = datetime.now()
         upcoming_tasks = [task for task in self.tasks if task.deadline > now and not task.completed]
         upcoming_tasks.sort(key=lambda task: task.deadline)
         return upcoming_tasks

    def save_tasks(self, filename="tasks.json"):
        task_data = [task.__dict__ for task in self.tasks]
        with open(filename, "w") as f:
            json.dump(task_data, f, default=str)

    def load_tasks(self, filename="tasks.json"):
        try:
            with open(filename, "r") as f:
                task_data = json.load(f)
                self.tasks = [Task(**data) for data in task_data]
        except FileNotFoundError:
            self.tasks = []

    def send_notification(self, title, message):
        plyer.notification.notify(title=title, message=message, timeout=10)

    def check_reminders(self):
        upcoming_tasks = self.get_upcoming_tasks()
        for task in upcoming_tasks:
             if task.deadline - datetime.now() <= timedelta(minutes=15):
                 self.send_notification(title="Task Reminder", message=f"Task {task.title} is due soon")

if __name__ == '__main__':
    task_manager = TaskManager()
    task_manager.load_tasks()

    
    task1 = Task(task_id=1, title="Grocery Shopping", description="Buy groceries for the week", priority="High", deadline=datetime(2025, 5, 27, 12, 0))
    task2 = Task(task_id=2, title="Book Appointment", description="Book a doctor appointment", priority="Medium", deadline=datetime(2025, 5, 28, 15, 0))
    task_manager.add_task(task1)
    task_manager.addounit(task2)

    schedule.every(1).minutes.do(task_manager.check_reminders)

    while True:
        schedule.run_pending()
        time.sleep(1)

File: question_1/task.kt
class Priority { LOW, MEDIUM, HIGH }
data class Task(
    val id: Int,
    var title: String,
    var description: String,
    var priority: Priority = Priority.MEDIUM,
    var deadline: Date,
    var isRecurring: Boolean = false,
    var recurrenceIntervalMillis: Long? = null // for recurring tasks
)
class TaskManager {
     val tasks = mutableMapOf<Int, Task>()
     val timers = mutableMapOf<Int, Timer>()
     var nextId = 1


    fun addTask(
        title: String,
        description: String? = null,
        priority: Priority = Priority.MEDIUM,
        deadline: Date? = null,
        isRecurring: Boolean = false,
        recurrenceIntervalMillis: Long? = null
    ): Task {
        val task = Task(nextId++, title, description, priority, deadline, isRecurring, recurrenceIntervalMillis)
        tasks[task.id] = task
        scheduleReminder(task)
        return task
    }

    
    fun editTask(
        id: Int,
        title: String,
        description: String,
        priority: Priority,
        deadline: Date,
        isRecurring: Boolean,
        recurrenceIntervalMillis: Long 
    ): Boolean {
        val task = tasks[id] ?: return false
        title?.let { task.title = it }
        description?.let { task.description = it }
        priority?.let { task.priority = it }
        deadline?.let { task.deadline = it }
        isRecurring?.let { task.isRecurring = it }
        recurrenceIntervalMillis?.let { task.recurrenceIntervalMillis = it }
        cancelReminder(id)
        scheduleReminder(task)
        return true
    }

    
    fun deleteTask(id: Int): Boolean {
        cancelReminder(id)
        return tasks.remove(id) != null
    }

    
    fun listTasks(): List<Task> = tasks.values.toList()


    fun scheduleReminder(task: Task) {
        task.deadline?.let { deadline ->
            val now = Date()
            val delay = deadline.time - now.time
            if (delay > 0) {
                val timer = Timer()
                if (task.isRecurring && task.recurrenceIntervalMillis != null) {
                    
                    timer.scheduleAtFixedRate(delay, task.recurrenceIntervalMillis!!) {
                        println("Reminder: Task '${task.title}' is due now or recurring.")
                    }
                } else {
                    
                    timer.schedule(delay) {
                        println("Reminder: Task '${task.title}' is due now.")
                    }
                }
                timers[task.id] = timer
            }
        }
    }

    
    fun cancelReminder(taskId: Int) {
        timers[taskId]?.cancel()
        timers.remove(taskId)
    }
}


fun main() {
    val taskManager = TaskManager()


    val deadline = Date(System.currentTimeMillis() + 10_000)
    val task1 = taskManager.addTask(
        title = "Submit report",
        priority = Priority.HIGH,
        deadline = deadline
    )

    
    val recurringDeadline = Date(System.currentTimeMillis() + 5_000)
    val task2 = taskManager.addTask(
        title = "Water plants",
        priority = Priority.LOW,
        deadline = recurringDeadline,
        isRecurring = true,
        recurrenceIntervalMillis = 15_000
    )

    
    taskManager.editTask(task1.id, description = "Sprint planning")

    
    taskManager.listTasks().forEach { println(it) }

    Thread.sleep(60_000)
}